Add 3rd method for computing meridional heat transport using velocities and temperature fields

dhruvbhagtani opened the pull request:
Closes #266.

navidcy commented:
@rmholmes could you review this?

rmholmes reviewed (changes requested):
This seemed to run fine for me, looks good. Well done @dhruvbhagtani 

However, please see my review comments. In particular, I think you need to add a bit more information on what these different estimates mean physically, and how they differ. Maybe no code changes are required, but more explanatory text certainly is.

Happy to discuss further if you have questions.

rmholmes commented on DocumentedExamples/Meridional_heat_transport.ipynb:
```diff
@@ -5,15 +5,19 @@
    "metadata": {},
```
<p>I suggest adding a sentence or two here indicating the accuracy of the different methods. E.g. <code>temp_yflux_adv_int_z</code> is an online diagnostic and thus is accurate (except of course that it only captures the advective components, not the parameterized components). <code>net_sfc_heating</code>  is an approximate calculation that only applies if the ocean is in steady state. Likewise, using <code>ty_trans</code>,<code>temp</code> is approximate as it neglects contributions from time varying correlations between transport and temperature at frequencies faster than the output frequency of the diagnostics themselves.</p><p> </p><p>You shouldn't need <code>dxt</code> and <code>dzt</code> for these calculations, since you're using the transport <code>ty_trans</code> not velocity.    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

dhruvbhagtani replied:
For the `temp_yflux_adv` MHT, when you say that it does not measure the parameterised components, it can also be understood that it doesn't capture the MHT due to bolus transport, which I think is also advective. So, isn't this statement better: `it only captures the advective components, not the parameterised diffusive components`?

rmholmes replied:
<p>Yes, hence you want to say "only captures resolved advective components, not the parameterised diffusive or advective components.".</p>

dhruvbhagtani replied:
Okay, I understand now, thanks!

rmholmes commented on DocumentedExamples/Meridional_heat_transport.ipynb:
```diff
@@ -5,15 +5,19 @@
    "metadata": {},
```
<p>Why are you not using dask and a distributed client?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

dhruvbhagtani replied:
I tried it with Dask before but it runs out of memory, possibly due to bad chunking. I realised that it might be another task to find a chunking style that works in this example, and then to explain chunking here as the notebook is already long. Do you think we should nevertheless have Dask here?

navidcy replied:
Yes. Include it. It's a good idea to ask for help though -- you don't need to figure everything yourself.

eg cc @angus-g, @dougiesquire

rmholmes commented on DocumentedExamples/Meridional_heat_transport.ipynb:
```diff
@@ -5,15 +5,19 @@
    "metadata": {},
```
<p>As described at https://github.com/COSIMA/access-om2/issues/139#issuecomment-639278547, for full accuracy we should also be including the frazil component of the surface heat flux via the diagnostic <code>frazil_3d_int_z</code>. If this diagnostic is not available for this run, please make a note of it.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

rmholmes replied:
<p>The actual value of the heat capacity used in ACCESS-OM2 is <code>Cp = 3992.10322329649</code>.  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

rmholmes replied:
<p>A comment here would be useful to describe how these observational estimates are derived. As far as I know, they are derived effectively using method 2 (but possibly with some tendency from OHC observations taken into account)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

rmholmes commented on DocumentedExamples/Meridional_heat_transport.ipynb:
```diff
@@ -5,15 +5,19 @@
    "metadata": {},
```
<p>Re "for this method to work" -&gt; You're referring here to the potential dependence of the meridional heat transport on the reference temperature associated with the fact that there could be a net meridional volume flux. In fact, both the above methods (the <code>temp_flux_adv_int_z</code> and <code>net_sfc_heating</code>) also do not account for net meridional volume transports. Thus their MHT will depend on the reference temperature. </p><p> </p><p>Hence, for this calculation here I would suggest first including a version which just calculates the net meridional heat transport without worrying about the reference temperature dependence. This is the only version of this calculation that can be properly compared to methods 1 and 2. Also, for this method you should not need to use <code>dxt</code> or <code>dzt</code> , only the transport diagnostic itself.</p><p> </p><p>If you wanted to, you could also include a method 4 then which is the current method 3 -&gt; i.e. what is the difference if you do subtract off the reference-temperature-dependent component of MHT associated with the net meridional volume transport? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

dhruvbhagtani replied:
I see your point. In that case, I think it might be useful to rewrite method 3 to be consistent with other methods, and then add a note that all of these methods are dependent on the temperature scale as there might be net meridional volume flux?

rmholmes replied:
<p>This comment still hasn't been addressed. I'd suggest just changing the wording at this point. In particular, you don't need the "For this method to work... southward flow". If you want to just compare to method 1, using <code>temp_flux_adv_int_z</code>,then you don't want to do this subtraction).</p><p> </p><p>I'd suggest also adding "mass" in between "meridional" and "transport" in the first sentence.</p><p> </p><p>EDIT: I see you have this extra text below. So I think you just missed removing it from here.</p>

rmholmes commented on DocumentedExamples/Meridional_heat_transport.ipynb:
```diff
@@ -5,15 +5,19 @@
    "metadata": {},
```
<p>As I mentioned above, there are other differences here. Specifically, the <code>shflux</code> method assumes steady state in the OHC. Also, your current 3rd method as calculated above is different because it extracts only the non-reference-temperature dependent component of MHT (which differs from the other calculations both through the potential tendency of sea level at different locations and due to the presence of net surface volume flux at different latitudes).</p><p> </p><p>With regards to the spatially uniform specific heat capacity - this is not an approximation. The model uses Conservative Temperature as its prognostic temperature variable which has a constant heat capacity associated with it (effectively, Conservative Temperature is defined such that you can use a constant specific heat capacity, i.e., it is "conservative"). If you're interested in more details on this please see Trevor's recent paper in GMD (https://gmd.copernicus.org/articles/14/6445/2021/)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

dhruvbhagtani replied:
Thanks for the reference, Ryan. I didn't know about it!

navidcy commented:
thanks for the review @rmholmes 

@dhruvbhagtani are you OK addressing these remarks at your convenience?

dhruvbhagtani commented:
Thanks for the review Ryan. Yes, I'll make the required changes in the next few days.

dhruvbhagtani commented:
I have added more text for easier readability. The one place I am not fully confident is in the observation dataset comment, particularly if all relevant datasets can be found in the webpage. I have rewritten method 3 to be consistent with other methods, and mentioned this detail in the beginning.

navidcy commented:
@dhruvbhagtani let me know when you are done or if you need help

dhruvbhagtani commented:
Sure Navid, I am happy to try a few things on my own first before asking for help. I aim to add dark support by the end of this week.

dhruvbhagtani commented:
This is done now :)

navidcy commented:
@dhruvbhagtani I see several comments by @rmholmes; are these resolved? (if so, can you mark them as resolved?)

navidcy commented:
@rmholmes any chance you can have another look at this?

navidcy commented:
I changed the environment to be the latest (so that the notebook doesn't go stale soon) and rerun all cells.
There are a bunch of warnings from xarray or dask urging us to rechunk

```python
/g/data/hh5/public/apps/miniconda3/envs/analysis3-23.07/lib/python3.10/site-packages/xarray/core/dataset.py:270: UserWarning: The specified chunks separate the stored chunks along dimension "yt_ocean" starting at index 540. This could degrade performance. Instead, consider rechunking after loading.
  warnings.warn(
/g/data/hh5/public/apps/miniconda3/envs/analysis3-23.07/lib/python3.10/site-packages/xarray/core/dataset.py:270: UserWarning: The specified chunks separate the stored chunks along dimension "yt_ocean" starting at index 540. This could degrade performance. Instead, consider rechunking after loading.
```

```python
/g/data/hh5/public/apps/miniconda3/envs/analysis3-23.07/lib/python3.10/site-packages/xarray/core/dataset.py:270: UserWarning: The specified chunks separate the stored chunks along dimension "xt_ocean" starting at index 288. This could degrade performance. Instead, consider rechunking after loading.
  warnings.warn(
/g/data/hh5/public/apps/miniconda3/envs/analysis3-23.07/lib/python3.10/site-packages/xarray/core/dataset.py:270: UserWarning: The specified chunks separate the stored chunks along dimension "xt_ocean" starting at index 288. This could degrade performance. Instead, consider rechunking after loading.
```

@dhruvbhagtani any idea? I trust that the warnings are there for a reason.
@dougiesquire, @angus-g, @micaeljtoliveira ?

dhruvbhagtani commented:
I have seen these warnings before with newer environments; not sure why they are, considering that we don't chunk any data explicitly.

navidcy commented:
> I have seen these warnings before with newer environments; not sure why they are, considering that we don't chunk any data explicitly.

Well they are trying to inform us… but I am not sure what they imply.

dougiesquire commented:
> I changed the environment to be the latest (so that the notebook doesn't go stale soon) and rerun all cells.
There are a bunch of warnings from xarray or dask urging us to rechunk

I don't seem to get these using `analysis3-unstable`. Which environment are you using?

navidcy commented:
23.07

dougiesquire commented:
Strange, I get no warnings with `23.07`...

dougiesquire commented:
The [`preferred_chunks` attribute was implemented for reading from NetCDF files in xarray>=v2023.09.0](https://github.com/pydata/xarray/pull/7948). This changed some of the behaviour of the `chunks` kwarg in `xarray.open_dataset`. I suspect the warnings here are related to this change, but it's difficult to say more without being able to reproduce.

navidcy commented:
@rmholmes the PR requires your approval to merge.

rmholmes commented:
Sorry this is on my list but I'm travelling this week. Will get to it Monday.

navidcy commented:
No worries!

rmholmes commented on DocumentedExamples/Meridional_heat_transport.ipynb:
```diff
@@ -5,15 +5,21 @@
    "metadata": {},
```
<p>Much better! My only remaining suggestions:</p><ul><li>In the "net_sfc_heating" section make it clear that it is approximate *because of* the steady state assumption. Also, I know <code>net_sfc_heating</code> is missing the frazil formation, but I think it gets everything else (at least for the current version of the code, as described in <a href="https://github.com/COSIMA/access-om2/issues/139" target="_blank">https://github.com/COSIMA/access-om2/issues/139</a>)? So I think you should remove the "land-ice heat discharges, and ice calving"? Finally, do you have <code>frazil_3d_int_z</code> for these runs? If so, you could just add it to <code>net_sfc_heating</code> and you've got everything. If not, then note specifically here that this diagnostic is needed.</li><li>I suggest changing "temperature flux at each latitude..." to "a reference-temperature dependent heat flux as there may be net meridional mass transports for which a unique heat transport cannot be defined (see <a href="see [Holmes et al. 2018](https://doi.org/10.1029/2019GL085160" target="_blank">Holmes et al. 2018</a> and <a href="see Holmes et al. 2018 and [Forget and Ferriera 2019](https://www.nature.com/articles/s41561-019-0333-7" target="_blank">Forget and Ferriera 2019</a> for a discussion)."</li><li>I'm not sure what you mean by the "To ensure conservation,..." sentence. Mass is still conserved - there's no issue there. I think you mean something like "To compute a reference-temperature independent heat transport, one would need to subtract the net meridional mass transport first."?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

rmholmes replied:
<p><s>Looks like this equation isn't displaying properly?</s></p><p> </p><p>-&gt; Actually, this is just a problem here in "ReviewNB". Looks like all is good in the actual notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

rmholmes replied:
<p>Would suggest changing "is then used" to "could be used", since you don't actually do it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

rmholmes replied:
<p>Can we just add a note here saying "These observations are derived using various methods, in particular using surface flux observations and method 2 (which assumes a steady state)".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/COSIMA/cosima-recipes/pull/285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

rmholmes commented:
Looks good to me @dhruvbhagtani. Almost there. Just a few minor text changes that I've suggested in the above review. Happy to approve once you've done those (should take 2 mins).

dhruvbhagtani commented:
All done now! Sorry for the delay, I messed up commits and majority of the time was spent in figuring out how to get things back.
